js/ui/paygUnlockDialog.js

```diff
@@ -0,0 +1,395 @@
+
+const CODE_ENTRY_REQUIRED_LENGTH = 8;
+
+var SPINNER_ICON_SIZE = 16;
+var SPINNER_ANIMATION_DELAY = 1.0;
+var SPINNER_ANIMATION_TIME = 0.3;
+
+var MESSAGE_FADE_OUT_ANIMATION_TIME = 0.5;
+
+var UnlockStatus = {
+    NOT_VERIFYING: 0,
+    VERIFYING: 1,
+    FAILED: 2,
+    SUCCEEDED: 3
+};
+
+var PaygUnlockCodeEntry = new Lang.Class({
+    Name: 'PaygUnlockCodeEntry',
+    Extends: St.Entry,
+    Signals: { 'code-changed' : { param_types: [GObject.TYPE_STRING] } },
+
+    _init: function(params) {
+        this.parent({ style_class: 'unlock-dialog-payg-entry',
+                      reactive: true,
+                      can_focus: true,
+                      x_align: Clutter.ActorAlign.FILL });
+
+        this._code = '';
+        this.clutter_text.ellipsize = Pango.EllipsizeMode.NONE;
+        this.clutter_text.x_align = Clutter.ActorAlign.CENTER;
+
+        this.connect('button-press-event', this._onButtonPressEvent.bind(this));
+
+        this.clutter_text.connect('captured-event', this._onCapturedEvent.bind(this));
+        this.clutter_text.connect('text-changed', this._onTextChanged.bind(this));
+    },
+
+    _onCapturedEvent: function(textActor, event) {
+        if (event.type() != Clutter.EventType.KEY_PRESS)
+            return Clutter.EVENT_PROPAGATE;
+
+        let character = event.get_key_unicode();
+
+        // We only support printable characters.
+        if (!GLib.unichar_isprint(character))
+            return Clutter.EVENT_PROPAGATE;
+
+        // Don't allow inserting more digits than required.
+        if (this._code.length >= CODE_ENTRY_REQUIRED_LENGTH)
+            return Clutter.EVENT_STOP;
+
+        // Allow digits only
+        if (GLib.unichar_isdigit(character))
+            this.clutter_text.insert_unichar(character);
+
+        return Clutter.EVENT_STOP;
+    },
+
+    _onTextChanged: function(textActor) {
+        this._code = textActor.text;
+        this.emit('code-changed', this._code);
+    },
+
+    _onButtonPressEvent: function() {
+        this.grab_key_focus();
+        return false;
+    },
+
+    addCharacter: function(character) {
+        this.clutter_text.insert_unichar(character);
+    },
+
+    setEnabled: function(value) {
+        this.reactive = value;
+        this.clutter_text.editable = value;
+    },
+
+    reset: function() {
+        this.text = '';
+    },
+
+    get code() {
+        return this._code;
+    },
+
+    get length() {
+        return this._code.length;
+    }
+});
+
+var PaygUnlockDialog = new Lang.Class({
+    Name: 'PaygUnlockDialog',
+
+    _init: function(parentActor) {
+        this._parentActor = parentActor;
+
+        this._entry = null;
+        this._errorMessage = null;
+        this._cancelButton = null;
+        this._nextButton = null;
+        this._spinner = null;
+        this._cancelled = false;
+
+        this._verificationStatus = UnlockStatus.NOT_VERIFYING;
+
+        // Clear the clipboard to make sure nothing can be copied into the entry.
```

Comment:
Will this clear the clipboard in the user’s session? I’d be pretty annoyed if I had something on the clipboard, then the computer locked, and then I lost what I had.